File: lib/Utils/lecture_colors.dart
import 'package:flutter/material.dart';
import 'FirebaseDatabaseService.dart';
import 'UnifiedDatabaseService.dart';

class LectureColors {
  // Permanent color cache - colors are cached forever once generated
  static Map<String, Color> _colorCache = {};
  static bool _isInitialized = false;

  /// Initialize colors for all existing tracking types (call once on app start)
  static Future<void> initializeColors() async {
    if (_isInitialized) return;
    
    try {
      final databaseService = FirebaseDatabaseService();
      final trackingTypes = await databaseService.fetchCustomTrackingTypes();
      
      // Generate and cache colors for all existing tracking types
      for (String trackingType in trackingTypes) {
        if (!_colorCache.containsKey(trackingType)) {
          _colorCache[trackingType] = _generateColorFromString(trackingType);
        }
      }
      
      _isInitialized = true;
    } catch (e) {
      // Even if fetch fails, mark as initialized to avoid repeated attempts
      _isInitialized = true;
    }
  }

  /// Get color for a lecture type - generates and caches if new
  static Future<Color> getLectureTypeColor(BuildContext context, String type) async {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;
    
    if (type.isEmpty) {
      return colorScheme.surfaceVariant.withOpacity(0.3);
    }
    
    // Check if color is already cached
    if (_colorCache.containsKey(type)) {
      return _colorCache[type]!;
    }
    
    // If not cached, this might be a new tracking type
    // Fetch latest tracking types to verify it exists
    try {
      final databaseService = FirebaseDatabaseService();
      final trackingTypes = await databaseService.fetchCustomTrackingTypes();
      
      if (trackingTypes.contains(type)) {
        // It's a valid tracking type - generate and cache color permanently
        final color = _generateColorFromString(type);
        _colorCache[type] = color;
        return color;
      } else {
        // Unknown type - return default color (don't cache)
        return colorScheme.surfaceVariant.withOpacity(0.3);
      }
    } catch (e) {
      // If fetch fails, generate color anyway (might be valid)
      final color = _generateColorFromString(type);
      _colorCache[type] = color;
      return color;
    }
  }

  /// Add color for a newly created tracking type
  static void cacheColorForNewType(String trackingType) {
    if (trackingType.isNotEmpty && !_colorCache.containsKey(trackingType)) {
      _colorCache[trackingType] = _generateColorFromString(trackingType);
    }
  }

  /// Get color synchronously if already cached, otherwise generate and cache
  static Color getLectureTypeColorSync(BuildContext context, String type) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;
    
    if (type.isEmpty) {
      return colorScheme.surfaceVariant.withOpacity(0.3);
    }
    
    // Return cached color if available
    if (_colorCache.containsKey(type)) {
      return _colorCache[type]!;
    }
    
    // Generate and cache color immediately for sync usage
    final color = _generateColorFromString(type);
    _colorCache[type] = color;
    return color;
  }

  /// Clear all cached colors (useful for testing)
  static void clearCache() {
    _colorCache.clear();
    _isInitialized = false;
  }

  /// Generate a consistent color from string input
  static Color generateColorFromString(String input) {
    return _generateColorFromString(input);
  }

  /// Generate a consistent color from string input
  static Color _generateColorFromString(String input) {
    final int hash = _customHash(input);
    // Generate more vibrant colors with better contrast
    final int r = ((hash & 0x0000FF) % 120) + 100;
    final int g = (((hash & 0xFF0000) >> 16) % 120) + 100;
    final int b = (((hash & 0x00FF00) >> 8) % 120) + 100;
    return Color.fromARGB(255, r, g, b);
  }

  /// Custom hash function for consistent color generation
  static int _customHash(String input) {
    int hash = 0;
    for (int i = 0; i < input.length; i++) {
      hash = 31 * hash + input.codeUnitAt(i);
    }
    return hash;
  }
}

File: android/app/src/main/kotlin/com/imnexerio/revix/LectureColors.kt
package com.imnexerio.revix

import android.content.Context
import android.graphics.Color
import android.content.SharedPreferences

/**
 * Android/Kotlin version of LectureColors with caching
 * Mirrors the logic from Flutter's lecture_colors.dart
 */
class LectureColors private constructor() {
    companion object {
        private const val CACHE_EXPIRY_MINUTES = 10
        private const val PREF_CACHE_PREFIX = "lecture_color_"
        private const val PREF_CACHE_TIME_PREFIX = "lecture_color_time_"
        
        /**
         * Get color for a lecture type with caching mechanism
         */
        fun getLectureTypeColor(context: Context, entryType: String): Int {
            if (entryType.isEmpty()) {
                return getDefaultTextColor()
            }
            
            val sharedPrefs = context.getSharedPreferences("LectureColorsCache", Context.MODE_PRIVATE)
            
            // Check cache first
            val cacheKey = PREF_CACHE_PREFIX + entryType
            val cacheTimeKey = PREF_CACHE_TIME_PREFIX + entryType
            val cachedColor = sharedPrefs.getInt(cacheKey, Int.MIN_VALUE)
            val cacheTime = sharedPrefs.getLong(cacheTimeKey, 0)
            
            val currentTime = System.currentTimeMillis()
            val cacheExpiryTime = CACHE_EXPIRY_MINUTES * 60 * 1000 // Convert to milliseconds
            
            // Return cached color if valid and not expired
            if (cachedColor != Int.MIN_VALUE && (currentTime - cacheTime) < cacheExpiryTime) {
                return cachedColor
            }
            
            // Generate new color for any non-empty entry_type (no validation needed)
            val color = generateColorFromString(entryType)
            
            // Cache the color
            sharedPrefs.edit()
                .putInt(cacheKey, color)
                .putLong(cacheTimeKey, currentTime)
                .apply()
                
            return color
        }
        
        /**
         * Get color synchronously if already cached, otherwise return default
         */
        fun getLectureTypeColorSync(context: Context, entryType: String): Int {
            if (entryType.isEmpty()) {
                return getDefaultTextColor()
            }
            
            val sharedPrefs = context.getSharedPreferences("LectureColorsCache", Context.MODE_PRIVATE)
            val cacheKey = PREF_CACHE_PREFIX + entryType
            val cachedColor = sharedPrefs.getInt(cacheKey, Int.MIN_VALUE)
            
            return if (cachedColor != Int.MIN_VALUE) {
                cachedColor
            } else {
                // Generate and cache immediately
                getLectureTypeColor(context, entryType)
            }
        }
        
        /**
         * Generate consistent color from string input (same logic as Flutter version)
         */
        private fun generateColorFromString(input: String): Int {
            val hash = customHash(input)
            // Generate more vibrant colors with better contrast (same logic as Flutter)
            val r = ((hash and 0x0000FF) % 120) + 100
            val g = (((hash and 0xFF0000) shr 16) % 120) + 100
            val b = (((hash and 0x00FF00) shr 8) % 120) + 100
            return Color.rgb(r, g, b)
        }
        
        /**
         * Custom hash function for consistent color generation (same as Flutter version)
         */
        private fun customHash(input: String): Int {
            var hash = 0
            for (char in input) {
                hash = 31 * hash + char.code
            }
            return hash
        }
        
        /**
         * Get default text color (system default)
         */
        private fun getDefaultTextColor(): Int {
            return Color.parseColor("#FF000000") // Default black text color
        }
        
        /**
         * Clear color cache (useful for testing or when tracking types change)
         */
        fun clearCache(context: Context) {
            val sharedPrefs = context.getSharedPreferences("LectureColorsCache", Context.MODE_PRIVATE)
            val editor = sharedPrefs.edit()
            
            // Remove all cached colors
            val allPrefs = sharedPrefs.all
            for ((key, _) in allPrefs) {
                if (key.startsWith(PREF_CACHE_PREFIX) || key.startsWith(PREF_CACHE_TIME_PREFIX)) {
                    editor.remove(key)
                }
            }
            editor.apply()
        }
    }
}
